fsthttp/imageopto/imageopto.go

```diff
@@ -0,0 +1,787 @@
+	Enable                             EnableOpt
+	Fit                                Fit
+	Format                             Format
+	Frame                              uint32
+	Height                             PixelsOrPercentage
+	Level                              Level
+	Metadata                           Metadata
+	Optimize                           OptimizeLevel
+	Orient                             Orientation
+	Pad                                *Sides
+	Precrop                            *Crop
+	Profile                            Profile
+	Quality                            uint32
+	ResizeFilter                       ResizeAlgorithm
+	Saturation                         int
+	Sharpen                            *Sharpen
+	Trim                               *Sides
+	TrimColor                          *TrimColor
+	Width                              PixelsOrPercentage
+}
+
+func encodeCommas(s string) string {
+	return strings.ReplaceAll(strings.ReplaceAll(s, ",", "%2C"), ":", "%3A")
+}
+
+func (o *Opts) validateParams() error {
+
+	// TODO(dgryski): not every parameter is validated for correctness here
+
+	if !o.Region.isSet() {
+		return errors.New("imageopto: region is not set")
+	}
+
+	if b := o.BgColor; b != nil {
+		if b.A < 0.0 || b.A > 1.0 {
+			return errors.New("imageopto: alpha out of range 0..1")
+		}
+	}
+
+	if b := o.Brightness; b != 0 {
+		if b < -100 || b > 100 {
+			return errors.New("imageopto: brightness out of range -100 .. 100")
+		}
+	}
+
+	if c := o.Contrast; c != 0 {
+		if c < -100 || c > 100 {
+			return errors.New("imageopto: contrast out of range -100 .. 100")
+		}
+	}
+
+	if c := o.Canvas; c != nil {
+		if err := c.validate(); err != nil {
+			return err
+		}
+	}
+
+	if c := o.Crop; c != nil {
+		if err := c.validate(); err != nil {
+			return err
+		}
+	}
+
+	if d := o.Dpr; d != 0 {
+		if d < 0 || d > 10 {
+			return errors.New("imageopto: dpr out of range 0 .. 10")
+		}
+	}
+
+	if h := o.Height; h.isSet() {
+		if err := h.validate(); err != nil {
+			return err
+
+		}
+	}
+
+	if p := o.Pad; p != nil {
+		if err := p.validate(); err != nil {
+			return err
+
+		}
+	}
+
+	if c := o.Precrop; c != nil {
+		if err := c.validate(); err != nil {
+			return err
+		}
+	}
+
+	if q := o.Quality; q != 0 {
+		if q > 100 {
+			return errors.New("imageopto: quality out of range 0 .. 10")
+		}
+	}
+
+	if s := o.Saturation; s != 0 {
+		if s < -100 || s > 100 {
+			return errors.New("imageopto: saturation out of range -100 .. 100")
+		}
+	}
+
+	if s := o.Sharpen; s != nil {
+		if err := s.validate(); err != nil {
+			return err
+		}
+	}
+
+	if t := o.Trim; t != nil {
+		if err := t.validate(); err != nil {
+			return err
+
+		}
+	}
+
+	if t := o.TrimColor; t != nil {
+		if err := t.validate(); err != nil {
+			return err
+		}
+	}
+
+	if w := o.Width; w.isSet() {
+		if err := w.validate(); err != nil {
+			return err
+
+		}
+	}
+
+	return nil
+}
+
+func (o *Opts) QueryString() string {
+	var args []string
```

Comment:
url.Query is no good here?

Reply:
Wanted to preserve ordering to make the test cases easier. (But yes technically I could.  Don't think it would be nicer though)